js/alert_banner.js

```diff
@@ -4,31 +4,108 @@
  */
 
 (function ($, Drupal, once) {
-
   'use strict';
 
-  // Function taken from https://github.com/onury/invert-color
-  function getLuminance(c) {
-    var i, x;
-    var a = []; // So we don't mutate.
-    for (i = 0; i < c.length; i++) {
-      x = c[i] / 255;
-      a[i] = x <= 0.03928 ? x / 12.92 : Math.pow((x + 0.055) / 1.055, 2.4);
+  /**
+   * Calculate relative luminance for color contrast.
+   * Function taken from https://github.com/onury/invert-color
+   *
+   * @param {string|Array} color - The color value (hex or RGB array).
+   * @return {number} The relative luminance value.
+   */
+  function getLuminance(color) {
+    let r, g, b;
+
+    // Handle RGB array
+    if (Array.isArray(color)) {
+      [r, g, b] = color.map(x => parseInt(x) / 255);
+    }
+    // Handle RGB string
+    else if (color.indexOf('rgb') >= 0) {
+      [r, g, b] = color.replace(/rgb\(|\)|\s/gi, '').split(',').map(x => parseInt(x) / 255);
+    }
+    // Handle hex color
+    else if (typeof color === 'string' && color.startsWith('#')) {
+      const hex = color.replace('#', '');
+      r = parseInt(hex.substr(0, 2), 16) / 255;
+      g = parseInt(hex.substr(2, 2), 16) / 255;
+      b = parseInt(hex.substr(4, 2), 16) / 255;
     }
+    // Default to black if color format is not recognized
+    else {
+      return 0;
+    }
+
+    // Calculate luminance using WCAG formula
+    const a = [r, g, b].map(x =>
+      x <= 0.03928 ? x / 12.92 : Math.pow((x + 0.055) / 1.055, 2.4)
+    );
+
     return 0.2126 * a[0] + 0.7152 * a[1] + 0.0722 * a[2];
   }
 
+  /**
+   * Parse date from HTML time element.
+   *
+   * @param {string} dateString - The HTML time element string.
+   * @return {Date|null} The parsed date or null if invalid.
+   */
+  function parseDateFromTimeElement(dateString) {
+    if (!dateString) {
+      return null;
+    }
+    const match = dateString.match(/datetime="([^"]+)"/);
+    if (match && match[1]) {
+      return new Date(match[1]);
+    }
+    return null;
+  }
+
+  /**
+   * Check if alert is within its date range.
+   *
+   * @param {Object} alert_item - The alert object.
+   * @return {boolean} Whether alert is within its date range.
+   */
+  function checkDateVisibility(alert_item) {
+    const now = new Date();
+
+    // Check start date
+    if (alert_item.field_alert_start_date) {
+      const startDate = parseDateFromTimeElement(alert_item.field_alert_start_date);
+      if (startDate && now < startDate) {
+        return false;
+      }
+    }
+
+    // Check end date
+    if (alert_item.field_alert_end_date) {
+      const endDate = parseDateFromTimeElement(alert_item.field_alert_end_date);
+      if (endDate && now > endDate) {
+        return false;
+      }
+    }
+
+    return true;
+  }
+
+  /**
+   * Check if alert should be visible on current page.
+   *
+   * @param {string} page_visibility_string - The page visibility string.
+   * @return {boolean} Whether alert should be visible.
+   */
```

Comment:
Good fix for page visibility. The regex replacement now properly handles wildcards. However, consider escaping special regex characters in the path:
```javascript
// Escape special regex characters except our wildcards
page_visibility = page_visibility.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
page_visibility = page_visibility.replace(/\\\*/g, '[^ ]*');
```